torchdata/stateful_dataloader/README.md

```diff
@@ -0,0 +1,215 @@
+your dataset has worker-specific state (eg RNG transform state) you can add `state_dict/load_state_dict` methods to your
+dataset.
+
+```py
+from typing import *
+import torch
+import torch.utils.data
+from torchdata.stateful_dataloader import StatefulDataLoader
+
+class MySampler(torch.utils.data.Sampler[int]):
+  def __init__(self, high: int, seed: int, limit: int):
+    self.seed, self.high, self.limit = seed, high, limit
+    self.g = torch.Generator()
+    self.g.manual_seed(self.seed)
+    self.i = 0
+
+  def __iter__(self):
+    while self.i < self.limit:
+      val = int(torch.randint(high=self.high, size=(1,), generator=self.g))
+      self.i += 1
+      yield val
+
+  def load_state_dict(self, state_dict: Dict[str, Any]):
+    self.i = state_dict["i"]
+    self.g.set_state(state_dict["rng"])
+
+  def state_dict(self) -> Dict[str, Any]:
+    return {"i": self.i, "rng": self.g.get_state()}
+
+# Optional: save dataset random transform state
+class NoisyRange(torch.utils.data.Dataset):
+  def __init__(self, high: int, mean: float, std: float):
+    self.high, self.mean, self.std = high, torch.tensor([float(mean)]), float(std)
+
+  def __len__(self):
+    return self.high
+
+  def __getitem__(self, idx: int) -> float:
+    if not (0 <= idx < self.high):
+      raise IndexError()
+    x = torch.normal(self.mean, self.std)
+    noise = x.item()
+    return idx + noise
+
+  def load_state_dict(self, state_dict):
+    torch.set_rng_state(state_dict["rng"])
+
+  def state_dict(self):
+    return {"rng": torch.get_rng_state()}
+
+for num_workers in [0, 2]:
+  print(f"{num_workers=}")
+  dl = StatefulDataLoader(NoisyRange(5, 1, 1), sampler=MySampler(5, 1, 10),
+      batch_size=2, drop_last=False, num_workers=num_workers)
+
+  batches = []
+  for i, batch in enumerate(dl):
+    batches.append(batch)
+    if i == 2:
+      sd = dl.state_dict()
+
+  dl.load_state_dict(sd)
+  batches2 = list(dl)
+
+  print(batches[3:])
+  print(batches2)
+
+"""
+Output:
+num_workers=0
+[tensor([-0.4526,  3.7948], dtype=torch.float64), tensor([6.5494, 3.0470], dtype=torch.float64)]
+[tensor([-0.4526,  3.7948], dtype=torch.float64), tensor([6.5494, 3.0470], dtype=torch.float64)]
+num_workers=2
+[tensor([3.7412, 1.2438], dtype=torch.float64), tensor([4.4807, 4.0036], dtype=torch.float64)]
+[tensor([3.7412, 1.2438], dtype=torch.float64), tensor([4.4807, 4.0036], dtype=torch.float64)]
+"""
+```
+
+## Saving Custom State with Iterable-Style Datasets
+
+Tracking iteration order with Iterable-style datasets requires state from each worker-level instance of the dataset to
+be captured. You can define `state_dict/load_state_dict` methods on your dataset which capture worker-level state.
+`StatefulDataLoader` will handle aggregation across workers and distribution back to the workers. Calling
+`load_state_dict` requires `StatefulDataLoader` to have same `num_workers` as those of the provided `state_dict`.
```

Comment:
`num_workers` need to be same even for map-style datasets correct? 

Separate discussions: 
i) should we consider relaxing that if map dataset doesn't have state?
ii) assert statement https://github.com/pytorch/data/blob/main/torchdata/stateful_dataloader/stateful_dataloader.py#L1211 can have an error message to make this check explicit for the user

Reply:
num_workers can be different for map-style if they only care about iteration order, I can clarify that